lib/pure/pegs.nim

```diff
@@ -85,6 +85,11 @@ type
     of pkBackRef..pkBackRefIgnoreStyle: index: range[0..MaxSubpatterns]
     else: sons: seq[Peg]
   NonTerminal* = ref NonTerminalObj
+  NtCallback* = object
+    enter*: proc(nt: NonTerminal, matchStart: int)
```

Comment:
These callbacks should be ``{.gcsafe.}``

Reply:
That's a tough one. When I declare the callbacks {.gcsafe.}, the compiler complains that "``rawMatch`` can have side effects", which is true, because the callbacks I make ``rawMatch`` call tend to have these. Btw. why didn't the compiler complain before I put the {.gcsafe.} there?

Reply:
Side effects are fine, but rawMatch is ``gcsafe`` so I don't understand this problem.

Reply:
Ah, all these procs are declared to be free of side effects. Mark the callbacks ``noSideEffect`` then?

Reply:
> Btw. why didn't the compiler complain before I put the {.gcsafe.} there?

Seems a compiler bug. :-(